File: frontend/src/api/referenceData.ts
import type { Script, NPC } from '@/types';
import type { PromptTemplate } from './templates';
import type { LLMConfig } from './llmConfigs';
import client from './client';

export interface ReferenceData {
  scripts: Script[];
  npcs: NPC[];
  templates: PromptTemplate[];
  llm_configs: LLMConfig[];
}

export const referenceDataApi = {
  /**
   * Fetch all reference data in a single request.
   * This reduces the number of API calls from 4+ to 1 for pages
   * that need scripts, npcs, templates, and llm_configs.
   */
  get: async (): Promise<ReferenceData> => {
    const response = await client.get('/reference-data');
    return response.data;
  },
};

File: backend/app/schemas/reference_data.py
"""Reference data schemas for aggregated API response."""

from pydantic import BaseModel, Field

from app.schemas.llm_config import LLMConfigResponse
from app.schemas.npc import NPCResponse
from app.schemas.prompt_template import TemplateResponse
from app.schemas.script import ScriptResponse


class ReferenceDataResponse(BaseModel):
    """Aggregated reference data response for frontend optimization."""

    scripts: list[ScriptResponse] = Field(default_factory=list, description="All scripts")
    npcs: list[NPCResponse] = Field(default_factory=list, description="All NPCs")
    templates: list[TemplateResponse] = Field(default_factory=list, description="All prompt templates")
    llm_configs: list[LLMConfigResponse] = Field(default_factory=list, description="All LLM configurations")
